src/module-config/module-config.ts

```diff
@@ -6,6 +6,16 @@ const configs: object[] = [];
 // An object with module names for keys and schemas for values.
 const schemas = {};
 
+// Pull default config file from module-config
+const configPromise = System.import("module-config");
+configPromise
+  .then(res => {
+    configs.unshift(res);
+  })
+  .catch(() => {
+    // do nothing; no config file has been provided via import map
```

Comment:
Hmm, swallowing this error would be rather confusing for someone who wants `config-file` to load and doesn't understand why it's working.

I would recommend doing the following instead:

```js
let hasConfigFile
try {
  System.resolve('config-file')
  hasConfigFile = true
catch {
  hasConfigFile = false
}

if (hasConfigFile) {
  System.import('config-file').then(module => {
    configs.unshift(module)
  })
}
```

^^ This way allows the error message to be propagated to the browser console when the loading of config-file loads because of a problem with the module itself.

Reply:
This is great, thanks Joel!

However, bad news: since I'm now trying to load config files using `System.import` and `System.import` is async, what's happening is that `getConfig()` is resolving with `provide`'d and default config content prior to the `import`ed config becoming available.

Going down this route will take the entire `config` object that `getConfig` produces into async territory.

What's the easiest that we can make this for module developers?

Make `getConfig` return a Promise, and suggest that module developers do something like
```jsx
let config = null;
getConfig("my-module")
.then(res => {
  config = res;
}).catch(err => {
  // complain loudly
});

...
<Img src={config ? config.logo.url : "/media/placeholder"} />
```
?

Is that too cumbersome? Is there a better way?

Reply:
Hmm if we make it async then I think we should incorporate fetching the module config into https://github.com/openmrs/openmrs-react-root-decorator. Or we could delay [calling singleSpa.start()](https://github.com/openmrs/openmrs-esm-root-config/blob/master/src/openmrs-esm-root-config.defaults.js#L8) until after the config is loaded. Or we could expose a react hook that relies on [Suspense](https://reactjs.org/docs/react-api.html#suspense) to make the component code still seem synchronous even when it's not. In all situations, I do not think it's reasonable to expect all react components to have to think about the async nature of retrieving the configuration.

Reply:
Brilliant. Yes, I agree that we should not make module developers think about its async nature, now that I know we don’t have to.

Of those three ideas, I like delaying singleSpa.start most, since it would be compatible with non-React libraries?

Reply:
yeah delaying singleSpa.start has that advantage but also means that the time until first render is delayed. No dom elements from the microfrontends are created until singleSpa.start() is called. I think I actually lean towards a hook for this. And create Angular/Vue/Svelte abstractions that do the same as needed.

Reply:
Also have you thought about what to do if someone calls `provide()` after some things have already rendered? Do we want to support that use case, or no? If so, I think we'd maybe want to use an observable or similar (event emitter) so that those using the config know that it has changed.

Reply:
Ok, yeah I agree that, if making singleSpa.start synchronous with the config from import map loading (or 404ing) would cause singleSpa.start to be delayed most of the time, we should probably not do that.

With hooks, would that mean that modules might render with the wrong/default config momentarily before receiving the right config?

I wonder if there’s a way to download `config-file`  at the same time as all the other modules are downloading, rather than when esm-module-config is executed. Something in the root config?

Other possibility— if making the config-file synchronously before singleSpa.start usually yields performance that is acceptable if not optimal, and if `System.resolve` will throw an error for a 404 extremely quickly, maybe it would be reasonable to say that those interested in the performance boost should override the root config and use `provide`.

Reply:
Regarding `provide`— I guess the use case would be fetching configs from e.g. a server endpoint?

Do you think we should support that use case?

Reply:
Is it possible to just have singlespa start after the login config is loaded? This could solve concerns of having to wait given this will be a small file

Reply:
Is it possible to just have singlespa start after the login config is loaded? This could solve concerns of having to wait given this will be a small file

Reply:
> With hooks, would that mean that modules might render with the wrong/default config momentarily before receiving the right config?

No. The idea is to change `getConfig()` to be asynchronous, but hide the asynchronicity in a react hook. Example code:

```js
// inside of application code
function Foo() {
  const moduleConfig = useModuleConfig()

  return <img src={moduleConfig.logoUrl} />
}

// inside esm-module-config
function useModuleConfig() {
  const [config, setConfig] = React.useState(null)
  const moduleName = React.useContext(ModuleNameContext)
  if (!config) {
    // throwing a promise is weird, but is how React suspense is triggered
    // React will prevent the Foo component from rendering until the promise resolves.
    throw getConfig(moduleName)
  } else {
    return config
  }
}

// openmrs-react-root-decorator would maybe render the context provider with the module name
// This allows users of `useModuleConfig()` to not have to pass in the name of the module they are in over and over again.
const ModuleNameContext = React.createContext()
```

Also see https://codepen.io/joeldenning/pen/ExxZEbj?editors=0011 for a demo of this working.

Reply:
Wow, hidden async... the future we live in is wild. Anyway, this seems beautiful and good. I'll give this a try. Thank you Joel.

Reply:
Awesome. How does ModuleNameContext get passed from the application code (the calling context) to esm-module-config? As written above, esm-module-config doesn't know what ModuleNameContext is supposed to be. React Context is totally new territory for me.